### DISCUSSION TITLE
Caching static assets (e.g. fonts)

### DISCUSSION THREAD

**Discussion by kwiat1990** [remix-run/remix#4389] [Q&A] [answered]
I have a hard time while trying to figure out how I can define a "Cache-Control" header with other cache strategy for all assets like fonts or images than the default one with of one hour caching.

I have seen this [page](https://remix.run/docs/en/v1/api/conventions#entryservertsx) in Remix docs and then this [MR] (https://github.com/remix-run/remix/commit/b1ce044f812a9bd47a1cffbc1e598bfb2a2e5868#diff-dd7f7c2ce19967ada5f6cad3cd3ddb29aad3b3ad2b877d6051c969a4a828152a) but it it still not clear to me if I can target such assets and set its cache strategy explicitly.

I run default Remix server, so I don't have any custom server file, which I could adjust.

**Comment by xHomu** [comment, marked as answer]
Remix Server is just a customized Express Server, and you can customize it after ejecting:

https://github.com/kiliman/rmx-cli#-eject-ras

This creates a server.js in your app root.  Then you can add

```jsx
// fonts are unlikely to change, so make them cache forever
app.use(
  "/fonts",
  express.static("public/fonts", { immutable: true, maxAge: "1y" })
);
```

**Reply by kwiat1990** [reply]
Thanks for pointing out this option. I didn't know it can be done this way.

One problem though: when I run `npm run dev` or newly created `dev:node` required env variables won't be loaded and the app crashes.

The scripts look following:
```json
    "build": "run-s build:*",
    "build:css": "npm run generate:css -- --minify",
    "build:remix": "remix build",
    "dev": "remix build && run-p \"dev:*\"",
    "dev:css": "npm run generate:css -- --watch",
    "dev:remix": "remix watch",
    "generate:css": "npx tailwindcss -i ./app/styles/tailwind.css -o ./app/tailwind.css",
    "start:locally": "dotenv remix-serve build",
    "start": "cross-env NODE_ENV=production node ./server.js",
    "dev:node": "cross-env NODE_ENV=development nodemon ./server.js --watch ./server.js"
```

**Reply by kwiat1990** [reply]
Well, that problem with missing environmental variables was because express doesn't load them automatically on its own. This must be made explicitly, for example with the help of `dotenv` package, which need be imported inside `server.js`:  

```js 
require("dotenv").config("./.env")
```

Alternatively, and it's my current approach, there's a package called `env-cmd` which takes care of loading `.env` file, so this `cross-env NODE_ENV=production node ./server.js` becomes `env-cmd node ./server.js` .
 
My mistake was to think that `cross-env` take care of that.

**Comment by felixh10r** [comment]
@kwiat1990 I've worked around this by including the assets in the app folder (e.g. /app/assets), importing them in /app/root.tsx, and referencing them in the jsx:

```tsx
import robotoLatinRegular from "~/assets/fonts/roboto-v20-latin-regular.woff2";
...
<style
      dangerouslySetInnerHTML={{
        __html: `@font-face {
  font-family: 'Roboto';
  ...
  src: local(''), url('${robotoLatinRegular}') format('woff2');
}`}} />
```

This way, the assets are included in the bundle, include a hash in the filename, and are served under /build with the correct headers.
This works similarly to other bundlers, but I didn't find this documented explicitly in the Remix docs 🤔
However, it would be very desirable to have a Remix way to set cache headers for assets in the /public directory.

**Reply by kwiat1990** [reply]
@felixh10r, I have tried your approach as I encountered problems with hot reloading and collecting data from my app and sending it to Metronome, while using express server. I do not quite like this solution because of added overhead (moving assets into `app` dir and building that script with font-faces) but at the end of the day it does work and it works well 😃 .
